File: typescript/src/model/ral_method.ts
import { Definition } from "model/definition";
import { Identity } from "model/identity";
import { ObjectRef } from "model/object_ref";
import { RalObject } from "model/ral_object";
import { RalObjectWithRole } from "model/ral_object_with_role";
import { SpecificProperties } from "model/specific_properties";
import { Template } from "model/template";


export class RalMethod {
    identity: Identity;
    definition: Definition;
    template: Template;
    methodState: string;
    specificProperties: SpecificProperties;
    //todo existenceStarts
    //todo duration
    inputObjects: RalObjectWithRole[];
    outputObjects: RalObject[];
    nestedMethods: RalMethod[];

    inputObjectsRef: ObjectRef[];
    outputObjectsRef: ObjectRef[];

    constructor(params: {
        identity: Identity,
        definition: Definition,
        template: Template,
        methodState: string,
        specificProperties: SpecificProperties,
        inputObjects: RalObjectWithRole[],
        outputObjects: RalObject[],
        nestedMethods: RalMethod[],
        inputObjectsRef: ObjectRef[],
        outputObjectsRef: ObjectRef[]
    }) {
        this.identity = params.identity;
        this.definition = params.definition;
        this.template = params.template;
        this.methodState = params.methodState;
        this.specificProperties = params.specificProperties;
        this.inputObjects = params.inputObjects;
        this.outputObjects = params.outputObjects;
        this.nestedMethods = params.nestedMethods;
        this.inputObjectsRef = params.inputObjectsRef;
        this.outputObjectsRef = params.outputObjectsRef;
    }

    toMap(): Record<string, any> {
        return {
            identity: this.identity.toMap(),
            definition: this.definition.toMap(),
            template: this.template.toMap(),
            methodState: this.methodState,
            specificProperties: this.specificProperties.toMaps(),
            inputObjects: this.inputObjects.map(inputObject => inputObject.toMap()),
            outputObjects: this.outputObjects.map(outputObject => outputObject.toMap()),
            nestedMethods: this.nestedMethods.map(nestedMethod => nestedMethod.toMap()),
            inputObjectsRef: this.inputObjectsRef.map(inputObjectRef => inputObjectRef.toMap()),
            outputObjectsRef: this.outputObjectsRef.map(outputObjectRef => outputObjectRef.toMap())
        }
    }

    static fromMap(map: Record<string, any>): RalMethod {
        const identity = Identity.fromMap(map.identity);
        const definition = Definition.fromMap(map.definition);
        const template = Template.fromMap(map.template);
        const methodState = map.methodState ?? "undefined";
        const specificProperties = SpecificProperties.fromMaps(map.specificProperties);

        let inputObjects: RalObjectWithRole[] = [];
        if (map.inputObjects != null) {
            inputObjects = map.inputObjects.map((inputObject: Record<string, any>) => RalObjectWithRole.fromMap(inputObject));
        }

        let outputObjects: RalObject[] = [];
        if (map.outputObjects != null) {
            outputObjects = map.outputObjects.map((outputObject: Record<string, any>) => RalObject.fromMap(outputObject));
        }

        let nestedMethods: RalMethod[] = [];
        if (map.nestedMethods != null) {
            nestedMethods = map.nestedMethods.map((nestedMethod: Record<string, any>) => RalMethod.fromMap(nestedMethod));
        }

        let inputObjectsRef: ObjectRef[] = [];
        if (map.inputObjectsRef != null) {
            inputObjectsRef = map.inputObjectsRef.map((inputObjectRef: Record<string, any>) => ObjectRef.fromMap(inputObjectRef));
        }

        let outputObjectsRef: ObjectRef[] = [];
        if (map.outputObjectsRef != null) {
            outputObjectsRef = map.outputObjectsRef.map((outputObjectRef: Record<string, any>) => ObjectRef.fromMap(outputObjectRef));
        }

        return new RalMethod({
            identity,
            definition,
            template,
            methodState,
            specificProperties,
            inputObjects,
            outputObjects,
            nestedMethods,
            inputObjectsRef,
            outputObjectsRef
        });
    }

}

File: dart/lib/src/model/ral_method.dart
import 'package:openral_core/src/model/definition.dart';
import 'package:openral_core/src/model/identity.dart';
import 'package:openral_core/src/model/object_ref.dart';
import 'package:openral_core/src/model/ral_object.dart';
import 'package:openral_core/src/model/ral_object_with_role.dart';
import 'package:openral_core/src/model/specific_properties.dart';
import 'package:openral_core/src/model/template.dart';

class RalMethod {
  Identity identity;
  Definition definition;
  Template template;
  String methodState;
  SpecificProperties specificProperties;
  List<RalObjectWithRole> inputObjects;
  List<RalObject> outputObjects;
  List<RalMethod> nestedMethods;
  List<ObjectRef> inputObjectsRef;
  List<ObjectRef> outputObjectsRef;

  RalMethod({
    required this.identity,
    required this.definition,
    required this.template,
    required this.methodState,
    required this.specificProperties,
    required this.inputObjects,
    required this.outputObjects,
    required this.nestedMethods,
    required this.inputObjectsRef,
    required this.outputObjectsRef,
  });

  Map<String, dynamic> toMap() {
    return {
      "identity": identity.toMap(),
      "definition": definition.toMap(),
      "template": template.toMap(),
      "methodState": methodState,
      "specificProperties": specificProperties.toMaps(),
      "inputObjects": inputObjects.map((inputObject) => inputObject.toMap()).toList(),
      "outputObjects": outputObjects.map((outputObject) => outputObject.toMap()).toList(),
      "nestedMethods": nestedMethods.map((nestedMethod) => nestedMethod.toMap()).toList(),
      "inputObjectsRef": inputObjectsRef.map((inputObjectRef) => inputObjectRef.toMap()).toList(),
      "outputObjectsRef": outputObjectsRef.map((outputObjectRef) => outputObjectRef.toMap()).toList(),
    };
  }

  static RalMethod fromMap(Map<String, dynamic> map) {
    final identityResult = Identity.fromMap(map["identity"]);
    final definitionResult = Definition.fromMap(map["definition"]);
    final templateResult = Template.fromMap(map["template"]);
    final methodState = map.containsKey("methodState") ? map["methodState"] : "undefined";
    final specificPropertiesResult = SpecificProperties.fromMaps(map["specificProperties"]);

    if (identityResult.isRight) {
      throw Exception("Identity parsing failed: ${identityResult.right}");
    }
    final identity = identityResult.left;

    if (definitionResult.isRight) {
      throw Exception("Definition parsing failed: ${definitionResult.right}");
    }
    final definition = definitionResult.left;

    if (templateResult.isRight) {
      throw Exception("Template parsing failed: ${templateResult.right}");
    }
    final template = templateResult.left;

    if (specificPropertiesResult.isRight) {
      throw Exception("SpecificProperties parsing failed: ${specificPropertiesResult.right}");
    }
    final specificProperties = specificPropertiesResult.left;

    List<RalObjectWithRole> inputObjects = [];
    if (map.containsKey("inputObjects")) {
      inputObjects = map["inputObjects"].map<RalObjectWithRole>((inputObject) => RalObjectWithRole.fromMap(inputObject)).toList();
    }

    List<RalObject> outputObjects = [];
    if (map.containsKey("outputObjects")) {
      outputObjects = map["outputObjects"].map<RalObject>((outputObject) => RalObject.fromMap(outputObject)).toList();
    }

    List<RalMethod> nestedMethods = [];
    if (map.containsKey("nestedMethods")) {
      nestedMethods = map["nestedMethods"].map<RalMethod>((nestedMethod) => RalMethod.fromMap(nestedMethod)).toList();
    }

    List<ObjectRef> inputObjectsRef = [];
    if (map.containsKey("inputObjectsRef")) {
      inputObjectsRef = map["inputObjectsRef"].map<ObjectRef>((inputObjectRef) => ObjectRef.fromMap(inputObjectRef)).toList();
    }

    List<ObjectRef> outputObjectsRef = [];
    if (map.containsKey("outputObjectsRef")) {
      outputObjectsRef = map["outputObjectsRef"].map<ObjectRef>((outputObjectRef) => ObjectRef.fromMap(outputObjectRef)).toList();
    }

    return RalMethod(
      identity: identity,
      definition: definition,
      template: template,
      methodState: methodState,
      specificProperties: specificProperties,
      inputObjects: inputObjects,
      outputObjects: outputObjects,
      nestedMethods: nestedMethods,
      inputObjectsRef: inputObjectsRef,
      outputObjectsRef: outputObjectsRef,
    );
  }
}
